### PR TITLE
Image + Ratio Component Fixes / Improvements

### PR DESCRIPTION
**Pull request by sghoweri**
## Jira
http://vjira2:8080/browse/BDS-1614 + http://vjira2:8080/browse/BDS-1555 (separate PR)

## Summary
Image-related bug fixes and performance improvements.

## Expected Performance Impact
- Decrease total # of image requests
- Decrease overall page size (~200KB decrease in size on reference page)
- Slight increase in Google lighthouse score on desktop, no significant change on mobile ([72](https://www.webpagetest.org/lighthouse.php?test=190723_MR_59ba4a66617f590b9ec0f43ead06f285&run=2) to [78](https://www.webpagetest.org/lighthouse.php?test=190723_JR_4968a013b2071575d6dcb757c98ee05a&run=3) on desktop, [58](https://www.webpagetest.org/lighthouse.php?test=190723_YB_5c5676ca7613a60fdbf9cdaf8eed379a&run=1) to [57](https://www.webpagetest.org/lighthouse.php?test=190723_E4_48682e09ac6b46ea3b7228aae23fbc9e&run=1) on 4G)

## Fixes + Updates
- Re-added `overflow: hidden` to the <bolt-ratio> component to prevent blurred / lazily loaded images from bleeding out of the container.
> Note: this particular style was actually already in place in older versions of Bolt but was mistakenly removed during some more recent refactor work: https://github.com/bolt-design-system/bolt/blob/v2.3.0/packages/components/bolt-ratio/src/ratio.scss#L13
- Fixed the value getting assigned to the main `` tag’s `src` attribute in the `<bolt-image>` Twig template. Previously this had been using the base64 encoded version of the image (when getting lazy loaded) but this was just adding extra HTML data without actually doing what it was supposed to be doing. Now this has been updated to specifically use the 1x1 placeholder pixel instead.
- Patched the lazysizes NPM library to automatically add the ability to optionally customize which elements lazysizes targets. This addresses the main issue with lazysizes not fully working the way it’s intended (ie. not defer loading images outside of the viewport + automatically setting the correct sizes prop to the rendered  tag) 
    - NOTE: as part of this library patch, the specific version of lazysizes installed from NPM has been locked down to significantly reduce the risk of this patch getting auto-applied.
- Updated the `<bolt-image>`'s Twig template to pass along any base64 placeholder + average image color prop data to the web component (if it exists). Fixes a visible flash that had been occurring when the web component initially boots up
- Updated the `<bolt-image>` JS logic to no longer re-blur the image once it has already loaded the first time. This update addresses a lazy-loaded related rendering quirk if the component's `src` / `srcset` props are updated after the image component has initially rendered.
- Updated the image component's JS to initially set the `sizes` prop to the component width when/if  the default `auto` value is set.
  - Also added the ability to recalculate this auto-set `sizes` prop when the window gets resized

### Misc
- Updated the `<bolt-image>` JS template to only add the `data-sizes` prop to the image component when lazyloading
- Updated the carousel Jest test to use 2 different image fixtures that are higher quality to help prevent unexpected VRT test failures
- Added a small CSS update to make sure low quality image placeholders are blurred as expected 
- Updated the default lazysizes config to disable preloading after the page has loaded to help reduce the total # of network requests

NOTES:
- We need to chat about the browser’s lookahead pre-parser logic (and how non-lazy loaded images will automatically load up the largest image size if an appropriate sizes prop isn’t specified)
- Automatically calculating the sizes prop really doesn’t seem to be possible if the image isn’t getting lazy loaded (due to the lookahead pre-parser) 

## How to test
Original (Unoptimized) Reference Page:
 https://boltdesignsystem.com/pattern-lab/patterns/04-pages-10-d8-product-pages-product-t2/04-pages-10-d8-product-pages-product-t2.html

Optimized Reference Page: 
https://boltdesignsystem-erc0tiwqq.now.sh/pattern-lab/patterns/04-pages-10-d8-product-pages-product-t2/04-pages-10-d8-product-pages-product-t2.html 

- [x] Benchmark the same reference page (above) using https://www.webpagetest.org/ to compare results
- [x] Confirm the `sizes` attribute is consistently getting set on rendered `` tags inside of `<bolt-image>`
- [ ] Review the Twig template for our T2 product page example as a reference for manually configuring most images on a page to automatically lazyload (except the ones above the fold)
- [ ] @danielamorse, also check out the preliminary support I added for customizing an image's vertical alignment. I just stubbed out the basics for this but this should largely be all that we need (sans demos and docs like we chatted about) to get something out the door

### CODE REVIEW

**Comment by sghoweri** [comment]
^ will get my Travis tests for this straightened out tomorrow morning — I suspect I mostly just need to update a couple code snapshots.

**Comment by sghoweri** [comment]
@danielamorse @remydenton sorry to bug you guys but any updates on this one?

File: packages/components/bolt-image/src/image.js [hunk 1]
```diff
@@ -13,6 +13,23 @@ import './_image-lazy-sizes';
 
 let cx = classNames.bind(imageStyles);
 
+let passiveIfSupported = false;
+
+// https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener#Improving_scrolling_performance_with_passive_listeners
+try {
```
Comment (danielamorse) [inline packages/components/bolt-image/src/image.js]: Ooh, if this works out I'd like to see it in `core`.

File: docs-site/src/pages/pattern-lab/_patterns/04-pages/10-d8-product-pages/product-t2.twig [hunk 2]
```diff
@@ -228,20 +231,16 @@
           color: "black",
           image: {
             src: "/images/content/screenshots/device-screenshot--tablet.jpg",
-            lazyload: false
+            lazyload: true,
```
[the remark this answers is not in the document]
Comment (sghoweri) [reply docs-site/src/pages/pattern-lab/_patterns/04-pages/10-d8-product-pages/product-t2.twig]: Correct! As a rule of thumb, most of the time we should lazyload images (default behavior) unless we know an image is in the critical path / super important to show up ASAP

File: docs-site/src/pages/pattern-lab/_patterns/04-pages/10-d8-product-pages/product-t2.twig [hunk 3]
```diff
@@ -228,20 +231,16 @@
...
+            _sizes: "(min-width: 1400px) 40.64285714vw, (min-width: 600px) 41vw, 72.5vw",
```
[the remark this answers is not in the document]
Comment (sghoweri) [reply docs-site/src/pages/pattern-lab/_patterns/04-pages/10-d8-product-pages/product-t2.twig]: lol. I knew this would pop up 😜

This pseudo comment was from me testing and retesting this page on Google Lighthouse to figure out the sweet spot on lazyloading vs letting the browser eagerly start to download this image if I tell it the approximate component size and behavior at certain breakpoints.

TLDR: for non-lazy loaded images, what would the “ideal” sizes prop be, given that a client-side JS solution would end up executing way too late?

The magic numbers here obviously aren’t something easily authorable by content authors, but perhaps something like Puppeteer could lazily crawl the site and tell Drupal what sizes to set... 🤷‍♂️

File: docs-site/src/pages/pattern-lab/_patterns/04-pages/10-d8-product-pages/product-t2.twig [hunk 4]
```diff
@@ -21,7 +21,10 @@
       contentItems: [{
         pattern: "image",
         src: "/images/content/backgrounds/background-tall-3.jpg",
-        lazyload: false
+        lazyload: false,
+        placeholder_color: "hsl(10, 14%, 8%)",
+        valign: "30%",
+        sizes: "100vw"
```
[the remark this answers is not in the document]
Comment (sghoweri) [reply docs-site/src/pages/pattern-lab/_patterns/04-pages/10-d8-product-pages/product-t2.twig]: I believe the browser already defaults to 100vw if you don’t specify a sizes attribute on vanilla  tags.

Regarding setting 100vw as the default on background images (if sizes isn’t manually set), I was thinking the exact same thing — I just didn’t have enough time to test making that update.

> Related question: what does sizes="auto" actually do? If the purpose of sizes is to give the browser a hint as to the actual width of the image, is auto ever necessary?

`sizes=“auto”` only really gets used if the image src and srcset is getting loaded via JS. 

Lazyloaded <bolt-image> use lazy sizes to calculate this. 

The main update I added is to have _non-lazyloaded_, web component-rendered images check if sizes wasn’t defined  and if it wasn’t, set it to the component’s current dimensions so the rendered <bolt-image> still grabs the best image size needed.

Does that make sense?

File: packages/components/bolt-image/src/image.js (hunk 1 above)
Comment (sghoweri) [reply packages/components/bolt-image/src/image.js]: Agreed!

File: packages/components/bolt-image/patches/lazysizes+5.1.0.patch [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+ 	var getWidth = function(elem, parent, width){
+ 		width = width || elem.offsetWidth;
+@@ -596,8 +600,21 @@
+ 			_: function(){
+ 				started = Date.now();
+
+-				lazysizes.elements = document.getElementsByClassName(lazySizesCfg.lazyClass);
+-				preloadElems = document.getElementsByClassName(lazySizesCfg.lazyClass + ' ' + lazySizesCfg.preloadClass);
++        lazysizes.elements = lazySizesCfg.getElements
```
[the remark this answers is not in the document]
Comment (sghoweri) [reply packages/components/bolt-image/patches/lazysizes+5.1.0.patch]: Thanks 😊

File: docs-site/src/pages/pattern-lab/_patterns/04-pages/10-d8-product-pages/product-t2.twig (hunk 4 above)
[the remark this answers is not in the document]
Comment (adamszalapski) [reply docs-site/src/pages/pattern-lab/_patterns/04-pages/10-d8-product-pages/product-t2.twig]: I have the same question as Dan but your explanation makes sense.

File: packages/build-tools/tasks/image-tasks.js [hunk 6]
```diff
@@ -107,7 +107,7 @@ async function writeImageManifest(imgManifest) {
   );
 }
 
-async function processImage(file, set, skipOptimization) {
+async function processImage(file, set, skipOptimization = false) {
```
Comment (adamszalapski) [inline packages/build-tools/tasks/image-tasks.js]: Maybe better will be changed `skipOptimization` to `optimization`?

**Comment by sghoweri** [comment]
@danielamorse thanks! See answers below to the last couple of questions I wasn't able to get around to yesterday:  

> - The performance results on the optimized page are ever-so-slightly worse that the original page. Speed index goes from 1.721s > 2.015s for example. Am I just not interpreting it properly? I do see fewer requests on the optimized page and the waterfall view generally looks a lot better.
Before: https://www.webpagetest.org/result/190805_Z7_fe5edf61ab7e28c86932c5198cf24cc1/
After: https://www.webpagetest.org/result/190805_X4_40b9c29c92b99c7c0f56c891036fee64/ 

This has been driving me up a wall to be honest 🙄

When I run the site through web.dev's Lighthouse tests I do see our scores jump from [around 58](https://lighthouse-dot-webdotdevsite.appspot.com/lh/html?url=https://boltdesignsystem.com/pattern-lab/patterns/04-pages-10-d8-product-pages-product-t2/04-pages-10-d8-product-pages-product-t2.html) to now be [around 67](https://lighthouse-dot-webdotdevsite.appspot.com/lh/html?url=https://boltdesignsystem-1d99057iq.now.sh/pattern-lab/patterns/04-pages-10-d8-product-pages-product-t2/04-pages-10-d8-product-pages-product-t2.html) (9 point improvement!).

My suspicion is that we're now hitting a bottleneck with our Javascript + lack of Critical CSS getting in the way hitting faster perf scores...

We should definitely keep an eye on this but I'm confident that our other [in-progress perf updates](https://www.webpagetest.org/video/compare.php?tests=190805_6S_6da4145e0a87c00caa2c834bcce375f1%2C190805_2Q_dca4a2d9d34a4fd771b3168849cf5891%2C190805_0T_d63508ed1da68924135d5a91c3387417&thumbSize=200&ival=100&end=visual) should help make this PR's improvements even more visible.  

> - How expensive is it to have a resize listener on every image for `sizes`? I see that it's sometimes offset by the passive event listener. Any concerns here?

TBD. I haven't noticed any major differences in hands on testing but we should definitly keep kicking the tires just in case. 

For what its worth, the passive event listener should help in the interim but I'd like to have us explore other potential alternatives like intersection observer / resize observer + do some deeper JS profiling as the next round of work on this.

> - The image component is super complex. For all the features we want I guess it has to be. Going forward, I'd love to seize any opportunity we have to simplify it so that it's more approachable to devs (myself included!).

100% agree. This is something we should put our heads together and to come up with some strategies to help manage these necessarily complex components (but in a way that's 

> - I love that *not* preloading lazy-loaded image is finally working. Big win for performance!
> - `valign` work looks great. I like that it has nice defaults and allows fine-grained control.
💯 thanks!

I still could use some help with the demos and packing up that feature but I figure this should at least help take care of all the big stuff 👍

File: packages/components/bolt-image/package.json [hunk 7]
```diff
@@ -30,10 +30,15 @@
...
     "tag": "next"
   },
+  "scripts": {
+    "postinstall": "patch-package"
+  },
   "dependencies": {
     "@bolt/components-ratio": "^2.5.3",
     "@bolt/core": "^2.5.3",
-    "lazysizes": "^4.1.4"
+    "lazysizes": "^5.1.0",
```
Comment (remydenton) [inline packages/components/bolt-image/package.json]: In the PR description, you say:

> as part of this library patch, the specific version of lazysizes installed from NPM has been locked down to significantly reduce the risk of this patch getting auto-applied.

Wise idea.  But bear with me here-- shouldn't that locking happen here in package.json by omitting the caret (e.g. `"lazysizes": "^5.1.0"`) rather than in yarn.lock?  That would allow downstream projects like Drupal to also get the locked version and apply the patch correctly.

Edit: I see you did have it locked at one point (d6ef4f3), changed by mistake?

File: packages/components/bolt-image/src/image.twig [hunk 8]
```diff
@@ -22,9 +22,9 @@
 {% set alt = alt ? alt : null %}
 
 {# Temp: prefix "ph" to preserve original value for web component, better pattern in the works #}
-{% set ph_color = placeholder_color|default(image_data_twig.color|default(schema.properties.placeholder_color.default)) %}
+{% set ph_color = placeholder_color|default(image_data_bolt.color|default(schema.properties.placeholder_color.default)) %}
 {# Note: image-specific base64 data is only generated in production mode, default data is used for local development #}
-{% set ph_image = placeholder_image|default(image_data_twig.base64|default(schema.properties.placeholder_image.default)) %}
+{% set ph_image = placeholder_image|default(image_data_bolt.base64|default(schema.properties.placeholder_image.default)) %}
```
Comment (remydenton) [inline packages/components/bolt-image/src/image.twig]: Just a question, I don't see a "base64" key for any of the images in `www/build/data/images.bolt.json`.  Should I?

File: packages/components/bolt-image/src/image.js (hunk 1 above)
Comment (sghoweri) [reply packages/components/bolt-image/src/image.js]: @remydenton I agree that it's kind of confusing... if you look at MDN's fine print, they do clarify that this reversal of behavior only applies to touch start and touch move events on the top level window / body elements.

> To prevent this problem, some browsers (specifically, Chrome and Firefox) have changed the default value of the passive option to true for the touchstart and touchmove events on the document-level nodes Window, Document, and Document.body.

End of the day, what really matters is that we're using code that works (see [Modernizr reference here](https://github.com/Modernizr/Modernizr/blob/master/feature-detects/dom/passiveeventlisteners.js)) + we get consistent results as expected in the browsers that support:

IE 11 🚫 
passive-ie11

Edge ✅
passive-edge

Firefox ✅ 
passive-firefox

Chrome ✅ 
passive-chrome

Safari ✅ 
passive-safari

File: packages/components/bolt-image/package.json (hunk 7 above)
Comment (sghoweri) [reply packages/components/bolt-image/package.json]: HMMM 🤔 

@remydenton good catch. I think this might have been a goof when merging in the latest from master. I'll get this updated 👍

File: packages/components/bolt-image/src/image.twig (hunk 8 above)
Comment (sghoweri) [reply packages/components/bolt-image/src/image.twig]: @remydenton that's prep for the next PR -- I ended up pulling out those base64 updates into #1247 to make sure this PR isn't THAT crazy 😄

File: packages/components/bolt-image/src/image.js (hunk 1 above)
Comment (remydenton) [reply packages/components/bolt-image/src/image.js]: Thanks-- that's all behavior I was expecting.  I really do think Mozilla has a typo then.  i.e.

> You can override this behavior by explicitly setting the value of `passive` to `false`, as shown here:

should be

> You can override this behavior by explicitly setting the value of `passive` to `true`, as shown here:

Not sure if there's any way to submit feedback on their docs though 🤔

Comment (remydenton) [reply packages/components/bolt-image/src/image.js]: To tie this off, on closer inspection the comment was right and the code was wrong.  I submitted a change to the MDN docs, which turned out to be altogether too easy...
